### ISSUE TITLE
Likely bug in Theano breaks Keras

### ISSUE THREAD

**Issue by the-moliver** [fchollet/keras#664]
It seems like there was a bug recently introduced in Theano that breaks parts of Keras in subtle ways, i.e. things run but don't work very well. The last good commit was:
https://github.com/Theano/Theano/commit/2595fea0688a6e08db00c9d0a349b51d820f9ab1
The code below works if you pull from this commit, but loss doesn't really decrease with bleeding edge Theano. If someone can confirm I'll report it to the Theano devs. I've only tried on GPU, so it may only apply there

```Python
from __future__ import print_function
import numpy as np
np.random.seed(1337)

from keras.utils.test_utils import get_test_data
from keras.models import Sequential
from keras.layers.core import TimeDistributedDense
import unittest

(X_train, y_train), (X_test, y_test) = get_test_data(nb_train=1000, nb_test=200, input_shape=(5, 10), output_shape=(5, 10),
                                                 classification=False)

model = Sequential()
model.add(TimeDistributedDense(X_train.shape[-1], y_train.shape[-1]))
model.compile(loss='hinge', optimizer='rmsprop')
history = model.fit(X_train, y_train, nb_epoch=12, batch_size=16, validation_data=(X_test, y_test), verbose=2)
```

**Comment by nehz** [comment]
Haven't checked it out, but the linked commit seems to only modify tests ?

**Comment by the-moliver** [comment]
Yup, the issue starts with the next commit: https://github.com/Theano/Theano/commit/41a8e89b63f5b4d4f9a0b942554ed5483e1ceab5

**Comment by cafaxo** [comment]
I can confirm this. I was getting NaN all over the place with LSTMs. I rolled back theano to 2595fea, and everything works as expected.

**Comment by fchollet** [comment]
Thanks for notifying us! We should definitely bring this up with the Theano dev. Keras is not doing anything particularly fancy with Theano, so if Keras is broken I assume a lot of other Theano-code must be broken as well.

**Comment by the-moliver** [comment]
Theano devs say it's now fixed, I'll test and update.

**Comment by the-moliver** [comment]
Looks like that did the trick. I would advise everyone to update to the newest Theano!

**Comment by fchollet** [comment]
Awesome, thanks!
